### PR TITLE
Add MutableChannelGroup and ChannelGroupMutator

### PR DESCRIPTION
**Pull request by LKJacky**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

This PR adds  MutableChannelGroup and ChannelGroupMutator and related modules. They are core modules for pruning

## Modification

1. Add MutableChannelGroup and ChannelGroupMutator
2. modify DynamicConv.convert_from, where gorup conv will not be convert to dynamic conv
3. update tests of graph

## Checklist

**Before PR**:

- [ ok] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ok ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ok] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/models/mutators/channel_mutator/channel_group_mutator.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+        Example:
+            {'net.3_(0, 16)_out_2_in_1': 16, 'net.0_(0, 8)_out_2_in_1': 8}
+        """
+        templabe = {}
+        for group in self.prunable_groups:
+            templabe[group.name] = group.current_choice
+        return templabe
+
+    def sample_structure(self) -> Dict[str, Union[int, float]]:
```
Comment (HIT-cwh) [inline mmrazor/models/mutators/channel_mutator/channel_group_mutator.py:96]: 这里方不方便统一下接口，sample_choices

File: mmrazor/models/mutators/channel_mutator/channel_group_mutator.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        template = self.structure_template()
+        for key in template:
+            template[key] = self._name2group[key].sample_choice()
+        return template
+
+    def apply_structure(self, config: Dict[str, Union[int, float]]):
```
Comment (HIT-cwh) [inline mmrazor/models/mutators/channel_mutator/channel_group_mutator.py:102]: 方不方便统一下接口，sample_choices

File: mmrazor/models/mutables/mutable_channel/mutable_channel_container.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        self.name = ''
+        self.num_channels = num_channels
+        self.mutable_masks: IndexDict[SimpleMutableChannel] = IndexDict()
+
+    # choice
+
+    @property
+    def current_choice(self):
+        if len(self.mutable_masks) == 0:
+            return torch.ones([self.num_channels]).bool()
```
Comment (HIT-cwh) [inline mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:26]: 初始化tensor的时候需要考虑下device

File: mmrazor/models/mutables/mutable_channel/channel_groups/mutable_channel_group.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        return name
+
+    @property
+    def is_prunable(self):
+        """bool: if the channel-group is prunable"""
+        have_dynamic_op = False
+        all_node_prunable = True
+        for channel in self.input_related + self.output_related:
+            if channel.node.is_prunable is False:
+                all_node_prunable = False
```
Comment (HIT-cwh) [inline mmrazor/models/mutables/mutable_channel/channel_groups/mutable_channel_group.py:45]: 可以提前break

File: mmrazor/models/mutables/mutable_channel/channel_groups/mutable_channel_group.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+            if is_dynamic_op(channel.module):
+                have_dynamic_op = True
```
Comment (HIT-cwh) [inline mmrazor/models/mutables/mutable_channel/channel_groups/mutable_channel_group.py:47]: 可以提前break

File: mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        self.current_choice = self._choice
+
+    # private methods
+    def _prepare_choices(self):
+        for choice in self.candidate_choices:
+            assert isinstance(choice, self.choice_type)
+        self.candidate_choices = sorted(self.candidate_choices)
+
+    def _generate_mask(self, num_activated_channels):
+        mask = torch.zeros([self.num_channels])
```
Comment (HIT-cwh) [inline mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py:69]: 考虑device

File: mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        return self.candidate_choices[-1]
+
+    # prepare
+    def prepare_for_pruning(self):
+        super().prepare_for_pruning()
+        self._choice = self.max_choice
+        self.current_choice = self._choice
+
+    # private methods
+    def _prepare_choices(self):
```
Comment (HIT-cwh) [inline mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py:63]: 私有方法用双下划线？

File: mmrazor/models/mutators/channel_mutator/channel_group_mutator.py (hunk 1 above)
Comment (LKJacky) [reply mmrazor/models/mutators/channel_mutator/channel_group_mutator.py]: sample_structure -> sample_subnet, as same as NAS

File: mmrazor/models/mutators/channel_mutator/channel_group_mutator.py (hunk 2 above)
Comment (LKJacky) [reply mmrazor/models/mutators/channel_mutator/channel_group_mutator.py]: structure -> subnet

File: mmrazor/models/mutables/mutable_channel/mutable_channel_container.py (hunk 3 above)
Comment (LKJacky) [reply mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:24]: there is no way to determine device used, as there is no other parameters.
the device will be changed when used in dynamic op.

File: mmrazor/models/mutators/channel_mutator/channel_group_mutator.py (hunk 1 above)
Comment (HIT-cwh) [reply mmrazor/models/mutators/channel_mutator/channel_group_mutator.py]: sample_choices is used in OneShotModuleMutator. (https://github.com/open-mmlab/mmrazor/blob/dev-1.x/mmrazor/models/mutators/module_mutator/one_shot_module_mutator.py#L54)

File: mmrazor/models/mutables/mutable_channel/mutable_channel_container.py (hunk 3 above)
Comment (HIT-cwh) [reply mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:24]: There is an attribute `mask` in class `SimpleMutableChannel` which can provide device informaction.

File: mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py (hunk 7 above)
Comment (LKJacky) [reply mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py:63]: it seems that we use single '_' according to other codes.

File: mmrazor/models/mutators/channel_mutator/channel_group_mutator.py (hunk 1 above)
Comment (LKJacky) [reply mmrazor/models/mutators/channel_mutator/channel_group_mutator.py]: get

File: mmrazor/models/mutables/mutable_channel/mutable_channel_container.py (hunk 3 above)
Comment (LKJacky) [reply mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:24]: Yes, but MutableChannelContainer doesn't have an attribute 'mask'.

File: mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py (hunk 7 above)
Comment (HIT-cwh) [reply mmrazor/models/mutables/mutable_channel/groups/one_shot_channel_group.py:63]: It seems that single '_' means protected method and double '_' means private method.

File: mmrazor/models/mutables/mutable_channel/mutable_channel_container.py (hunk 3 above)
Comment (HIT-cwh) [reply mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:24]: 如果没有device信息，会不会导致device不同报错呢

File: mmrazor/models/mutators/channel_mutator/channel_group_mutator.py [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+        def is_dynamic_op(module, module_name):
+            """determine if a module is a dynamic op for fx tracer."""
+            return isinstance(module, ChannelDynamicOP)
+
+        self.group_class.prepare_model(supernet)
+        self._name2module = dict(supernet.named_modules())
+
+        if self.tracer_cfg['type'] == 'BackwardTracer':
+            self.tracer: BackwardTracer = TASK_UTILS.build(self.tracer_cfg)
```
[the remark this answers is not in the document]
Comment (LKJacky) [reply mmrazor/models/mutators/channel_mutator/channel_group_mutator.py]: Not all tracer needs building, as i put it in prepare_from_supernet.
I move buiding into ModuleGraph.init_using_backward_tracer

File: mmrazor/models/mutables/mutable_channel/mutable_channel_container.py (hunk 3 above)
Comment (LKJacky) [reply mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:24]: No, mask is used in dynamicop, and dynamicop always transfer the mask to its using device.

Comment (HIT-cwh) [reply mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:24]: 不好意思我忘记了，之前的mutable也没关注device的事

File: mmrazor/models/mutators/channel_mutator/base_channel_mutator.py [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        elif issubclass(group_cfg, MutableChannelGroup):
+            group_class = group_cfg
+            group_args = {}
+        else:
+            raise NotImplementedError()
+        return group_class, group_args
+
+    # implementation of abstract functions
+
+    def search_groups(self) -> Dict:
```
Comment (HIT-cwh) [inline mmrazor/models/mutators/channel_mutator/base_channel_mutator.py:131]: `@property`

File: mmrazor/models/mutators/channel_mutator/base_channel_mutator.py [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+    def current_choices(self):
+        config = self.choice_template
+        for group in self.prunable_groups:
+            config[group.name] = group.current_choice
+        return config
+
+    # group manage
+
+    @property
+    def prunable_groups(self) -> List[MUTABLECHANNELGROUP]:
```
Comment (HIT-cwh) [inline mmrazor/models/mutators/channel_mutator/base_channel_mutator.py:114]: What's the difference between `prunable_groups` and `search_groups`.

File: mmrazor/models/mutables/mutable_channel/mutable_channel_container.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+from mmrazor.registry import MODELS
+from mmrazor.utils import IndexDict
+from .base_mutable_channel import BaseMutableChannel
+from .simple_mutable_channel import SimpleMutableChannel
+
+
+@MODELS.register_module()
+class MutableChannelContainer(BaseMutableChannel):
+
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutables/mutable_channel/mutable_channel_container.py:14]: Add necessary Notes and update docstring

File: mmrazor/models/mutables/mutable_channel/simple_mutable_channel.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+from mmrazor.registry import MODELS
+from ..derived_mutable import DerivedMutable
+from .base_mutable_channel import BaseMutableChannel
+
+
+@MODELS.register_module()
+class SimpleMutableChannel(BaseMutableChannel):
+
+    def __init__(self, num_channels, **kwargs) -> None:
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutables/mutable_channel/simple_mutable_channel.py:14]: Add necessary Notes and update docstring

File: mmrazor/models/mutables/mutable_channel/stack_mutable_channel.py [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+import torch
+
+from mmrazor.registry import MODELS
+from .base_mutable_channel import BaseMutableChannel
+
+
+@MODELS.register_module()
+class StackMutableChannel(BaseMutableChannel):
+
+    def __init__(self, num_channels: int, **kwargs):
```
Comment (sunnyxiaohu) [inline mmrazor/models/mutables/mutable_channel/stack_mutable_channel.py:11]: Add necessary Notes and update docstring
